### PR TITLE
Begin production rule definition

### PR DESCRIPTION
**Pull request by OR13**
The intention of the Pull Request is to simply define the existing production rules in greater detail, this PR is supposed to be editorial, I am not trying to change any of the existing behavior.

***
<a href="https://pr-preview.s3.amazonaws.com/w3c/vc-jwt/pull/11.html" title="Last updated on Oct 25, 2022, 1:53 PM UTC (d503d99)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/vc-jwt/11/bfcd55e...d503d99.html" title="Last updated on Oct 25, 2022, 1:53 PM UTC (d503d99)">Diff</a>

### CODE REVIEW

File: index.html
Comment (Sakurann) [inline index.html:559]: Assuming this text will be removed later on... but breaking changes themselves are not the goal if they are needed to meet the goal of clarifying production rules, we will introduce them ;)

```suggestion
        There are currently 2 competing solutions to this problem. It is a goal
        of the v2 work to resolve them and
        reduce production rules to a single, simple, set of instructions that
        any implementer can easily meet if they poss a software library
        supporting [[RFC7515]] or [[RFC7519]].
```

File: index.html
Comment (Sakurann) [inline index.html:569]: Can we add an explanation why do these members need to be translated to the equivalents of the JWT claims?
iss, nbf, all of them are optional in RFC 7519 itself. They only become required when the JWT is used for a specific purpose, and the only place I can think of right now is an ID Token.

File: index.html
Comment (Sakurann) [inline index.html:582]: which members are added to the header based on the content of the Credential..?

File: index.html
Comment (OR13) [reply index.html:569]: hmm, yes, we should do that... but I am not sure we need to do that now... My goal was to write the production rules for all interpretations of v1 first... Can I convert this comment to an issue for discussion?

File: index.html
Comment (Sakurann) [inline index.html:606]: why define this logic if it is not a top-level claim `iss`?

Comment (OR13) [reply index.html:606]: because you may be mapping from:

```
issuer: { id, name } -> payload.iss === id
```

and you don't want someone to do this:

```
issuer: { id, name } -> payload.iss === { id, name }
```

^ AFAIK, this is illegal regarding iss:
 
 
 >  The "iss" (issuer) claim identifies the principal that issued the
 >   JWT.  The processing of this claim is generally application specific.
 >   The "iss" value is a case-sensitive string containing a StringOrURI
 >   value.  Use of this claim is OPTIONAL.
   
 https://www.rfc-editor.org/rfc/rfc7519.html#section-4.1.1

File: index.html
Comment (OR13) [reply index.html:582]: afaik, none right now. 

But since the header is singed over, we must define complete construction of it to satisfy production rules.

I suspect we will define some additional "security meta data" that is also an argument of product rules.... `privateKeyJwk` being an example... which might include `kid` and `alg` which would go in the header.

File: index.html
Comment (Sakurann) [reply index.html:569]: yes.

Comment (OR13) [reply index.html:569]: https://github.com/w3c/vc-jwt/issues/15#issue-1374577341

File: index.html
Comment (TallTed) [reply index.html:606]: I think `principal` (at least as used in JWT specs) is closer to `agent` or `entity` than it is to `subject`.

Comment (OR13) [reply index.html:606]: interesting... @selfissued @Sakurann @tplooker what do ya'll think?

File: index.html
Comment (selfissued) [inline index.html:569]: For a signed JWT, the payload of the JWS is called the "JWT Claims Set" in the JWT spec.  It would be better to use that term than "payload", I believe.

File: index.html
Comment (selfissued) [inline index.html:575]: Again "payload" -> "JWT Claims Set".  Please apply this correction wherever else it makes sense.

File: index.html
Comment (selfissued) [inline index.html:580]: Shouldn't the header always include at least the "alg" value used to sign with and probably the "kid"?  We're including those, not because of the contents of the credential, but as part of securing the JWT.

File: index.html
Comment (selfissued) [reply index.html:606]: Yes, "principle" in the JWT spec is a generic term like "entity".  We use it in the description of the "iss", "sub", and "aud" claims.  I don't think its use would be particularly useful in this spec, unless perhaps in usage parallel to that in the JWT spec.

File: index.html
Comment (selfissued) [inline index.html:610]: Nope.  issuanceDate -> iss.

File: index.html
Comment (selfissued) [inline index.html:619]: Don't mention "nbf" except as an optional claim to be present only when it's different than "iss".

File: index.html
Comment (selfissued) [inline index.html:642]: I don't understand what the issuance section is trying to convey.

File: index.html
Comment (selfissued) [inline index.html:648]: I don't understand what the "Instead of" section is trying to convey.

File: index.html
Comment (selfissued) [inline index.html:652]: We need Privacy Considerations text, but that can be a separate issue.

**Review by selfissued — CHANGES_REQUESTED** [review]
Several suggestions - some normative - some editorial.

**Comment by iherman** [comment]
The issue was discussed in a [meeting](https://www.w3.org/2017/vc/WG/Meetings/Minutes/2022-10-19-vcwg#section2-6) on 2022-10-19

- no resolutions were taken

<details><summary><i>View the transcript</i></summary>

#### 2.6. Begin production rule definition (pr vc-jwt#11)

_See github pull request [vc-jwt#11](https://github.com/w3c/vc-jwt/pull/11)._




**Orie Steele:** there are changes requested on this.  
… not sure when FPWD will be ready.  
… please submit all your comments ASAP.  

</details>

File: index.html
Comment (TallTed) [inline index.html:610]:
```suggestion
          <h4>issuanceDate -> iss</h4>
```

File: index.html
Comment (OR13) [reply index.html]: I've made this update in https://github.com/w3c/vc-jwt/pull/11/commits/b4c079b13d7d42840c2cd40d8502cdf5103c1b6d

The only section I did not update has to do with the mapping algorithm, which is harder to read and understand when not using the word payload. If its still a blocking requirement I prefer to file an issue to discuss and refine in a seperate PR.

File: index.html
Comment (OR13) [reply index.html:576]: https://github.com/w3c/vc-jwt/pull/11/commits/b4c079b13d7d42840c2cd40d8502cdf5103c1b6d

File: index.html
Comment (OR13) [reply index.html:581]: Good question. At this point, its not known what the header will be!

We add the `alg` based on the key being used to protect the header and payload.

In most implementations I have seen, this happens either automatically at the time of signing, or just before it, based on the private key or signer passed to the library.... Because `alg` is optional in the JWK spec, sometimes it needs to be passed as an additional argument, next to the private key... for example:

```js
const jwt = await new jose.SignJWT({ 'urn:example:claim': true })
  .setProtectedHeader({ alg: 'ES256' }) // required here, since we don't know what type of key will be used.
  .setIssuedAt()
  .setIssuer('urn:example:issuer')
  .setAudience('urn:example:audience')
  .setExpirationTime('2h')
  .sign(privateKey) // alg might not be known here, this is just a generic private key... not a JWK.

console.log(jwt)
```

https://github.com/panva/jose/blob/HEAD/docs/classes/jwt_sign.SignJWT.md#readme

File: index.html
Comment (OR13) [reply index.html:620]: see comment here: https://github.com/w3c/vc-jwt/pull/11#discussion_r1001770219

File: index.html
Comment (OR13) [reply index.html:643]: Happy to explain verbally.

This section closes the loop for readers regarding how a `header` and `payload` become a JWT.

We can't just magically assume that people know how to sign them properly.

Generally, the process is:

```
signature = base64url ( 
    rawSign ( 
       base64url (header) + '.' + base64url (payload) 
   )
)
```

See https://github.com/panva/jose/blob/8b9955515f6c81b322fe8603201c204de0bfbec2/src/jws/flattened/sign.ts#L129

File: index.html
Comment (OR13) [reply index.html:649]: There are 2 paths that are allowed, in v1.1:

1.  instead of...
2. in addition to...

This pull request only attempts to define `in addition to...` it's my opinion that `instead of ` should be defined as non conformant in v2.

See this thread: https://lists.w3.org/Archives/Public/public-credentials/2022Sep/0215.html

Regardless or the WG decision on this point, it's not my intention to define both paths in this PR.

File: index.html
Comment (OR13) [reply index.html:611]: I won't be accepting this change suggestion.

**Comment by OR13** [comment]
@selfissued I have addressed your feedback, let me know if you have additional concerns.

File: index.html
Comment (OR13) [reply index.html:611]: See also the current spec text: 
Screen Shot 2022-10-21 at 8 27 31 AM

File: index.html
Comment (OR13) [reply index.html:607]: I think there is a general mapping between `entity` and `issuer`, `holder` and `verifier`... thats probably best discussed on issues, please continue discussion here: https://github.com/w3c/vc-jwt/issues/35

File: index.html
Comment (TallTed) [reply index.html:649]: +1, deprecate `instead of`, retain `in addition to`

File: index.html
Comment (Sakurann) [reply index.html]: let's file an issue and discuss - consistency is usually better, but not always

File: index.html
Comment (David-Chadwick) [reply index.html:611]: I agree that given the current W3C DM definition, then issuanceDate maps to nbf

File: index.html
Comment (David-Chadwick) [reply index.html:649]: I can live with either but not both!. So we should decide in this standard which we are going to employ and mandate that, regardless of what the VC DM v2 spec says.

File: index.html
Comment (David-Chadwick) [reply index.html:611]: However, I also think that the spec should allow the signer to optionally add iat as well, especially if the nbf is not equal to "now", which iat will always be.

File: index.html
Comment (OR13) [reply index.html:649]: IMO, `instead of` is the one that should be removed... it creates a lot of implementation complexity, and it potentially destroys the usability of the data model, but thats not what this PR is about.

File: index.html
Comment (TallTed) [inline index.html:570]:
```suggestion
          We refer to the JWT Claims Set as <code>payload</code> in this section.
```

File: index.html
Comment (OR13) [reply index.html]: https://github.com/w3c/vc-jwt/issues/36

File: index.html
Comment (OR13) [reply index.html:611]: I prefer we take that conversation to issues, the intention of this PR is not to change anything, only to make what is currently legal, clearer.

**Comment by OR13** [comment]
I will need to update this based on the resolutions for `validFrom` and `validUntil`.

**Comment by iherman** [comment]
The issue was discussed in a [meeting](https://www.w3.org/2017/vc/WG/Meetings/Minutes/2022-10-26-vcwg#section1-5) on 2022-10-26

- no resolutions were taken

<details><summary><i>View the transcript</i></summary>

#### 1.5. Begin production rule definition (pr vc-jwt#11)

_See github pull request [vc-jwt#11](https://github.com/w3c/vc-jwt/pull/11)._




**Orie Steele:** That had comments / changes requested by Mike Jones, I believe I addressed them..  
… I did request a re-review. TallTed had some editorial comments that I accepted, I am just requesting Mike Jones to re-review..  
… That's it for VC-JWT PRs..  

**Michael Jones:** Do you want to talk about the issues?.  

**Orie Steele:** I was just doing PRs..  

**Michael Jones:** Thank you, that's fine, that's to TallTed for the recent approval..  

</details>

File: index.html
Comment (selfissued) [reply index.html:611]: Per the consensus call on last week's call, I believe that issuanceDate is going to become validFrom.  I'm fine with validFrom being mapped to `nbf`.

File: index.html
Comment (selfissued) [reply index.html:643]: All right.  This is innocuous.  I just wasn't sure that it's necessary.

File: index.html
Comment (selfissued) [reply index.html:649]: It's a more natural JWT representation to use exclusively JWT claims to represent information for which there is a defined JWT representation.  As such, `instead of` seems more natural.

**Review by selfissued — APPROVED** [review]
We may need to beef up the language about inclusion of `alg` as a required header parameter, but we could do that in a separate PR.

File: index.html
Comment (OR13) [reply index.html:649]: Lets continue the discussion on https://github.com/w3c/vc-jwt/issues/8

File: index.html
Comment (OR13) [reply index.html:643]: I think its very necessary actually, especially related to `kid` guidance and hints on discovering the publicKey... but perhaps we can follow up on an issue, lets use: https://github.com/w3c/vc-jwt/issues/30

File: index.html
Comment (OR13) [reply index.html:611]: Filed, https://github.com/w3c/vc-jwt/issues/38 We will not forget to do this.

**Comment by OR13** [comment]
I believe I have addressed all comments, I'd like to merge soon. @selfissued would you be willing to press the button?

**Comment by iherman** [comment]
The issue was discussed in a [meeting](https://www.w3.org/2017/vc/WG/Meetings/Minutes/2022-11-02-vcwg#section1-5) on 2022-11-02

- no resolutions were taken

<details><summary><i>View the transcript</i></summary>

#### 1.5. Begin production rule definition (pr vc-jwt#11)

_See github pull request [vc-jwt#11](https://github.com/w3c/vc-jwt/pull/11)._




**Orie Steele:** VC-JWT still has PR 11 open.  
… mike jones request changes and i have responded.  
… but there are still changes requested but i believe that those have been addressed.  
… but i would like to accelerate the pace with improvements to that draft.  
… additional reviews would help.  
… that is it for VC-JWT.  

**Michael Jones:** given that poll taken on previous call, there seems to be validFrom and validUntil.  
… language needs to be updated in the PR.  
… when to be merged in the vc data model.  

> *Michael Prorock:* +1 wait for upstream merge.

**Orie Steele:** intention was not make any changes, only document what was already required.  

> *Michael Prorock:* then correct / update.

**Orie Steele:** i don't thing we should coming with those changes in the PR.  

**Michael Jones:** ok, then i think i can approve.  

</details>

**Comment by OR13** [comment]
Open since august... only 1 approval... perhaps we should pause this work item?

**Comment by OR13** [comment]
I'm withdrawing this PR, until the WG decides they want to focus on this work item.

**Comment by iherman** [comment]
The issue was discussed in a [meeting](https://www.w3.org/2017/vc/WG/Meetings/Minutes/2022-11-30-vcwg#section2-4) on 2022-11-30

- no resolutions were taken

<details><summary><i>View the transcript</i></summary>

#### 2.4. Begin production rule definition (pr vc-jwt#11)

_See github pull request [vc-jwt#11](https://github.com/w3c/vc-jwt/pull/11)._




**Orie Steele:** has only one approval -- could indicate general lack of interest and focus on other issues - perhaps pause?.  

> *Manu Sporny:* +1 for pausing VC-JWT work item..

> *Manu Sporny:* (given the recent discussion around how we're going to process Work Items).

> *Dmitri Zagidulin:* I am somewhat surprised at the lack of engagement on that item..

> *Kerri Lemoie:* what is ivan's comment in reference to?.

> *Manu Sporny:* and seeing how VC-JWT duplicates work done in JsonWebSignature2020.

> *Orie Steele:* There has not been an FPWD for vc-jwt, and I don't recommend we seek one at this time..

> *Dave Longley:* it is true that VC-JWT is duplicative of JsonWebSignature2020.

> *Manu Sporny:* I would expect multiple -1s for VC-JWT based on engagement and how we're planning to adopt work items now..

**David Chadwick:** lots of interest in JWT- like at JFF plugfest. Lack of time but also implementors are getting it to work ok without so many errors as thought..  

> *Manu Sporny:* There were 17 different organizations implementing the DI approaches... vs. 6 for VC-JWT -- useful information on each cryptosuite..

**David Chadwick:** what is the difference between a credential and a verifiable credential- add topic to future agenda please?.  

**Kristina Yasuda:** chairs will discuss how to resolve that.  

> *Joe Andrieu:* +1 to discuss David's question. It's a good one for the group to clear up..

**Michael Jones:** agree with DavidC that there's a lot of interest in vc-jwt but doesn't indicate that there aren't any problems.  

**Orie Steele:** issues that remains: 1) mappings to the core data model, 2) concrete documentation on how to define how to obtain a public key to verify a credential..  

> *David Chadwick:* Orie: [did:jwk](did:jwk) solves the public key problem.

**Orie Steele:** should tackle core-data model issues first.  
… gain wg consensus about core-data model first..  

> *Sten Reijers:* +1 to Orie comment that VC-JWT is in a bad shape.

> *David Chadwick:* orie: the issue that Joe and I are discussing about credential vs verifiable credential is pertinent to jwt-vc.

> *Orie Steele:* its not that JWS/JWT or JOSE is in bad shape... its just W3C "versions" of it..

**Manu Sporny:** agree in pursuing consensus on JWT but new process could negatively affect this work. Implementors could indicate what they support..  
… agree to hit pause until the wg can focus on fixing the core data model concerns then comeback to vc-jwt.  

> *Orie Steele:* This is an "unofficial" draft for a vc-jws... which is massively simpler than vc-jwt: [https://transmute-industries.github.io/vc-jws/](https://transmute-industries.github.io/vc-jws/).

> *Sten Reijers:* Is there a link to VC-JOSE issue?.


> *Orie Steele:* Here is the original draft pre IETF 114: [https://or13.github.io/draft-osteele-vc-jose/](https://or13.github.io/draft-osteele-vc-jose/).

> *Orie Steele:* Here is the issue regarding vc-jose proposal: [https://github.com/w3c/vc-data-model/issues/971](https://github.com/w3c/vc-data-model/issues/971).

**Kristina Yasuda:** This work can happen within the process. Thank you for raising this topic. Chairs will discuss..  

</details>

**Comment by David-Chadwick** [comment]
I think the whole section
```
For backward compatibility with JWT processors, the following registered JWT claim names MUST be used, instead of or in addition to, their respective standard verifiable credential counterparts:

exp MUST represent the expirationDate property , encoded as a UNIX timestamp ( NumericDate ).
iss MUST represent the issuer property of a verifiable credential or the holder property of a verifiable presentation.
nbf MUST represent issuanceDate , encoded as a UNIX timestamp ( NumericDate ).
jti MUST represent the id property of the verifiable credential or verifiable presentation .
sub MUST represent the id property contained in the credentialSubject .
```
should be removed and the new JWT-VC spec should create the JWT independently of the contents of the VC or VP. In other words the signer of the JWT does not need to be the issuer of the VC but is whoever the signing entity is. Similarly the expiry date of the JWT is independent of the expirationDate of the credential. I think the new rules should be as follows

exp MUST represent the expiration date of the JWT (and is independent of the expirationDate of the VC/VP)
iss MUST represent the issuer of the JWT (and is independent of the issuer of the VC/VP)
nbf MUST represent the time the JWT is issued (i.e. now)
jti MUST represent the id of the JWT (and is independent of the id property of the VC/VP) 

sub is the most difficult one to decide. First VCs do not need to contain a subject.id property, so the current text cannot always apply. So we could say something like "sub MUST represent the subject of the JWT. Normally this will be the subject identifier of the VC or the holder identifier of the VP, but it could be a local identifier used by the issuer to identify the subject". Comments?

**Comment by OR13** [comment]
@David-Chadwick indeed, v2 can make breaking changes... thats not the intention of this PR.

This PR attempts to show exactly how broken v1.1 is, so that we can decide where to break from it for V2.

**Comment by OR13** [comment]
@brentzundel @Sakurann can we please merge this PR.

**Comment by David-Chadwick** [comment]
> This PR attempts to show exactly how broken v1.1 is, so that we can decide where to break from it for V2.

Now you have confused me. Surely this document is for the v2 DM only? If so then why are we creating v1.1 PRs in it? If you want clarifying text for v1.1 then you should do a PR on that document shouldn't you?
I think this PR should be dropped and we should only have production rules for V2 in it. (But we might want an Appendix detailing the differences between V2 and V1.1)

**Review by selfissued — APPROVED** [review]
I agree that we should merge this PR at this point.  I just re-reviewed it and it is purely editorial.

**Pull request merged**
